src/plugins/tonegen/indicator.c

```diff
@@ -0,0 +1,275 @@
+/*************************************************************************
+This file is part of tone-generator
+
+Copyright (C) 2010 Nokia Corporation.
+
+This library is free software; you can redistribute
+it and/or modify it under the terms of the GNU Lesser General Public
+License as published by the Free Software Foundation
+version 2.1 of the License.
+
+This library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
+USA.
+*************************************************************************/
+
+#define _GNU_SOURCE
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include <ngf/log.h>
+#include <trace/trace.h>
+
+#include "ausrv.h"
+#include "stream.h"
+#include "tone.h"
+#include "indicator.h"
+#include "dtmf.h"
+
+#define MAX_TONE_LENGTH (1 * 60 * 1000000)
+
+/* Special case for shorter tones, because we don't want to keep a silent
+ * stream around for a long time after playing the actual tone. */
+#define MAX_SHORT_TONE_LENGTH (1 * 5 * 1000000)
+
+#define TRACE(f, args...) trace_write(trctx, trflags, trkeys, f, ##args)
+#define LOG_CAT "tonegen-indicator: "
+
+static char     *ind_stream = STREAM_INDICATOR;
+static int       standard   = STD_CEPT;
+static void     *ind_props  = NULL;
+static uint32_t  vol_scale  = 100;
+
+int indicator_init(void)
+{
+    return 0;
+}
+
+
+void indicator_play(struct ausrv *ausrv, int type, uint32_t vol, int dur)
+{
+    struct stream *stream  = stream_find(ausrv, ind_stream);
+    uint32_t       timeout = dur ? dur : MAX_TONE_LENGTH;
+    
+    if (stream != NULL) {
+        dtmf_stop(ausrv);
+        indicator_stop(ausrv, PRESERVE_STREAM);
+    }
+    else {
+        stream = stream_create(ausrv, ind_stream, NULL, 0,
+                               tone_write_callback,
+                               tone_destroy_callback,
+                               ind_props,
+                               NULL);
+
+        if (stream == NULL) {
+            N_ERROR(LOG_CAT "%s(): Can't create stream", __FUNCTION__);
+            return;
+        }
+    }
+
+    vol = (vol_scale * vol) / 100;
+    
+    switch (type) {
```

Comment:
Obviously I don't understand all the different standards and tone types, so I assume that block just works as-is (it has worked before in tonegen after all).